## **NBA Betting Model**

### **The goal here is to see if we can create a base NBA betting model**

#### **First we will get the 2023 NBA Season Data**

##### Import libaries used

In [147]:
import requests
import pandas as pd
import json
import numpy as np
import warnings
warnings.filterwarnings("ignore")

In [25]:
def game_logs(season):
    # season: '2022-23'
    url = 'https://stats.nba.com/stats/leaguegamelog'
    headers= {'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36', 'Referer': 'https://www.nba.com/'}
    payload = {
        'Counter': '1000',
        'DateFrom': '',
        'DateTo': '',
        'Direction': 'DESC',
        'LeagueID': '00',
        'PlayerOrTeam': 'T',
        'Season': season,
        'SeasonType': 'Regular Season',
        'Sorter': 'DATE'}
    
    jsonData = requests.get(url, headers=headers, params=payload).json()
    
    rows = jsonData['resultSets'][0]['rowSet']
    columns = jsonData['resultSets'][0]['headers']
    
    df = pd.DataFrame(rows, columns=columns)
    
    df['DATE'] = pd.to_datetime(df['GAME_DATE'])
    df['bookie_date'] = df['DATE'].dt.strftime('%Y%m%d')
    df['home'] = np.where(df['MATCHUP'].str.contains('@'), 0, 1)
    del df['SEASON_ID']
    
    df['bet_date_id'] = df['TEAM_ABBREVIATION'] + df['bookie_date']
    
    return df 

#### Here will get the 2023 NBA season 

In [26]:
nba_games = game_logs('2022-23')    

In [28]:
nba_games.head()

TEAM_ID TEAM_ABBREVIATION  ... home  bet_date_id
0  1610612764               WAS  ...    1  WAS20230409
1  1610612756               PHX  ...    1  PHX20230409
2  1610612757               POR  ...    1  POR20230409
3  1610612758               SAC  ...    0  SAC20230409
4  1610612737               ATL  ...    0  ATL20230409

[5 rows x 32 columns]

### **Now we will get the betting odds**

##### **The site we are scrapping the data from is https://www.actionnetwork.com/nba/odds and it has the odds from the most commom bookies, Draftking, MGM, Fanduel ect..**


In [29]:
def book_odds(date):
    
    '''
    date => eg;20221018
    '''
    url = 'https://api.actionnetwork.com/web/v1/scoreboard/nba?bookIds=15,30,76,75,123,69,68,972,71,247,79&date=' + date
    
    headers = {'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:101.0) Gecko/20100101 Firefox/101.0'}
    
    r = requests.get(url, headers=headers)

    tb = json.loads(r.content.decode('utf-8'))
    
    games = tb['games']
    
    game_odds = []
    
    for i in games:
        odds = i['odds']
        
        dfs = []
        
        for o in odds:
            dfs.append(pd.DataFrame([o]))
        df = pd.concat(dfs)    
        
        t1 = pd.DataFrame([i['teams'][0]])
        t2 = pd.DataFrame([i['teams'][1]])
    
        t1 = t1[['id','full_name','display_name','short_name','location','abbr', 'standings']].add_suffix('_home')
        t2 = t2[['id','full_name','display_name','short_name','location','abbr', 'standings']].add_suffix('_away')
        
        t = pd.concat([t1,t2], axis = 1)
        dt =pd.concat([t,df], axis = 1)
        
        game_odds.append(dt)    
        
    df = pd.concat(game_odds)      
    
    book_ids = {15:'open', 30: 'best odds', 76:'Pointsbet',75:'betmgm',  123:'caesars', 69:'fanduel', 68:'draftking', 972:'betrivers', 71: 'playsugarhouse', 274:'unibet', 79: 'bet365' }
    
    df['bookie'] = df['book_id'].map(book_ids)
    
    df['bookie_date'] = date
    
    df['bet_date_id'] = df['abbr_home'] +  df['bookie_date']

    return df  


In [30]:
bet_odds = book_odds('20221018')

#### **Here we will get the betting odds for each team that played on that date** 

In [31]:
bet_odds.head()

id_home  full_name_home display_name_home  ... bookie bookie_date  bet_date_id
0      164  Boston Celtics           Celtics  ...   open    20221018  BOS20221018
0      164  Boston Celtics           Celtics  ...   open    20221018  BOS20221018
0      164  Boston Celtics           Celtics  ...   open    20221018  BOS20221018
0      164  Boston Celtics           Celtics  ...   open    20221018  BOS20221018
0      164  Boston Celtics           Celtics  ...   open    20221018  BOS20221018

[5 rows x 51 columns]

#### **The next step is get the betting odds for all the games of the season**

###### **We already have a column that has the beting dates in the nba_games dataframe, we then use it to get a list of those betting dates**

In [32]:
bookie_dates = list(set(nba_games['bookie_date'].tolist()))

##### **Now we will use bookie_dates list to run a loop on book_odds function, it will return a list of betting odds for each date**

In [33]:
# List where will store the betting odds dataframe results
nba_game_odds = []

# List to store what dates had an error and what the error was
error_notes = []

n = 0

# for each bookie date in the list bookie_dates, we will run the function book_odds
for i in bookie_dates:
    try:
        # Attempt some operation that might raise an error
        nba_game_odds.append(book_odds(i))  # This will raise a ValueError for non-integer elements

    except ValueError as e:
        # If an error occurs, catch it, add a note, and continue to the next element
        error_notes.append([i,e])
        print(e)  # to note any errors
        continue
        
    n = n + 1
    print(n)  # to note how far we are along the loop
    

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
No objects to concatenate
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
No objects to concatenate
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162


##### **Now we will check to see the size of nba_game_odds and error_notes lists, with the hope have having are error's list being empty**

In [34]:
print(len(nba_game_odds))
print(len(error_notes))

162
2


###### **We will now rerun the two errors we have again and update it to the nba_game_odds**  

In [41]:
if len(error_notes) > 0:
    for i in error_notes:
        nba_game_odds.append(book_odds(i[0]))

##### **Now we will concat our list of dataframes into one**

In [43]:
game_odds = pd.concat(nba_game_odds)
game_odds.sample(5)

id_home   full_name_home  ... bookie_date  bet_date_id
0      168   Indiana Pacers  ...    20230203  IND20230203
0      172  Milwaukee Bucks  ...    20230304  MIL20230304
0      168   Indiana Pacers  ...    20221214  IND20221214
0      164   Boston Celtics  ...    20230404  BOS20230404
0      174   Denver Nuggets  ...    20221227  DEN20221227

[5 rows x 51 columns]

#### **Now that we have the NBA betting odds, we are going to use betting odds set by the market as a feature for our nba_games dataframe**

##### **We will use the best money line odds as our feature**

In [44]:
nba_game_odds = game_odds[(game_odds['book_id'] == 30) & (game_odds['type'] == 'game')] #filter for best odds

In [45]:
nba_game_odds.head()

id_home     full_name_home  ... bookie_date  bet_date_id
0      161      Orlando Magic  ...    20230406  ORL20230406
0      160         Miami Heat  ...    20230406  MIA20230406
0      181  San Antonio Spurs  ...    20230406  SAS20230406
0      173          Utah Jazz  ...    20230406  UTA20230406
0      174     Denver Nuggets  ...    20230406  DEN20230406

[5 rows x 51 columns]

#### **We will now create ID columns so that we can map the moneyline to nba_games**

In [60]:
nba_game_odds['home_date_id'] = nba_game_odds['abbr_home'] + nba_game_odds['bookie_date']
nba_game_odds['away_date_id'] = nba_game_odds['abbr_away'] + nba_game_odds['bookie_date']

#### **We will now create a dictionary for moneyline and map it**

In [59]:
home_dic_ml = dict(zip(nba_game_odds.home_date_id, nba_game_odds.ml_home))
away_dic_ml = dict(zip(nba_game_odds.away_date_id, nba_game_odds.ml_away))
ml_dic = {**home_dic_ml, **away_dic_ml}
nba_games['ml_line'] = nba_games['bet_date_id'].map(ml_dic)

#### **We can see we have the moneyline column to our Nba games dataframe, called ml_line**

In [57]:
nba_games.head()

TEAM_ID TEAM_ABBREVIATION  ...  bet_date_id ml_line
0  1610612764               WAS  ...  WAS20230409    -130
1  1610612756               PHX  ...  PHX20230409    -900
2  1610612757               POR  ...  POR20230409   -1800
3  1610612758               SAC  ...  SAC20230409     130
4  1610612737               ATL  ...  ATL20230409    -260

[5 rows x 33 columns]

##### **Now we sort the columns to have a clean dataframe where string columns are next to each other.**

In [122]:
move_columns = ['TEAM_ABBREVIATION','TEAM_NAME','GAME_DATE','MATCHUP','GAME_ID','bookie_date','DATE', 'bet_date_id', 'MIN','TEAM_ID','home', 'ml_line','WL' ]
df = nba_games[move_columns + [col for col in nba_games.columns if col not in move_columns]]
df = df.drop(columns= ['VIDEO_AVAILABLE'])

In [123]:
df.head(5)

TEAM_ABBREVIATION               TEAM_NAME   GAME_DATE  ...  PF  PTS PLUS_MINUS
0               WAS      Washington Wizards  2023-04-09  ...  21  109         -5
1               PHX            Phoenix Suns  2023-04-09  ...  21  114         -5
2               POR  Portland Trail Blazers  2023-04-09  ...   9  101        -56
3               SAC        Sacramento Kings  2023-04-09  ...  18   95        -14
4               ATL           Atlanta Hawks  2023-04-09  ...  12  114         -6

[5 rows x 32 columns]

In [124]:
df['WL'] = np.where(df['WL'] == 'W', 1, 0)  #numericate WL where wins are 1 and 0 for loses
df = df.sort_values(by='DATE') #sort by date

#### **We will add the rolling averages for all the statistical columns by team** 

In [125]:
dt = df.join(df.groupby('TEAM_ID')[list(df)[12:]].expanding().mean().reset_index(level=0, drop=True).groupby(df['TEAM_ID']).shift().add_prefix('Rolling_Avg_'))

In [126]:
dt.head(50)

TEAM_ABBREVIATION  ... Rolling_Avg_PLUS_MINUS
2459               PHI  ...                    NaN
2456               LAL  ...                    NaN
2458               BOS  ...                    NaN
2457               GSW  ...                    NaN
2455               ORL  ...                    NaN
2432               WAS  ...                    NaN
2433               HOU  ...                    NaN
2434               ATL  ...                    NaN
2435               DET  ...                    NaN
2437               CHI  ...                    NaN
2438               MIA  ...                    NaN
2439               MEM  ...                    NaN
2440               NYK  ...                    NaN
2441               SAS  ...                    NaN
2442               TOR  ...                    NaN
2436               NOP  ...                    NaN
2444               OKC  ...                    NaN
2445               MIN  ...                    NaN
2446               CHA  ...                    NaN
2447               DEN  ...                    NaN
2448               PHX  ...                    NaN
2449               DAL  ...                    NaN
2454               UTA  ...                    NaN
2450               POR  ...                    NaN
2451               SAC  ...                    NaN
2452               IND  ...                    NaN
2453               BKN  ...                    NaN
2443               CLE  ...                    NaN
2428               LAL  ...                  -14.0
2430               PHI  ...                   -9.0
2431               MIL  ...                    NaN
2429               LAC  ...                    NaN
2414               NOP  ...                   22.0
2406               NYK  ...                   -3.0
2407               MEM  ...                    3.0
2408               BKN  ...                  -22.0
2409               GSW  ...                   14.0
2410               PHX  ...                    2.0
2411               MIN  ...                    7.0
2412               TOR  ...                    3.0
2413               POR  ...                    7.0
2415               CHI  ...                    8.0
2418               ATL  ...                   10.0
2417               WAS  ...                    7.0
2416               ORL  ...                   -4.0
2425               IND  ...                   -7.0
2424               HOU  ...                  -10.0
2423               SAS  ...                  -27.0
2426               CHA  ...                   27.0
2421               BOS  ...                    9.0

[50 rows x 52 columns]

#### **Now we will prepare the dataframe to run a Logistic Regression model**

In [137]:
# Select all numerical columns
data = dt.select_dtypes(include=[np.number])
data.dropna(inplace=True) #drop NAN values
data.head()

MIN     TEAM_ID  ...  Rolling_Avg_PTS  Rolling_Avg_PLUS_MINUS
2428  240  1610612747  ...            109.0                   -14.0
2430  240  1610612755  ...            117.0                    -9.0
2414  240  1610612740  ...            130.0                    22.0
2406  240  1610612752  ...            112.0                    -3.0
2407  240  1610612763  ...            115.0                     3.0

[5 rows x 44 columns]

##### **We will now set up X (feature dataframe) and y (prediction dataframe)**

In [141]:
X = data[list(data)[24:] + ['home', 'ml_line']]
y = data['WL']

In [145]:
# Import libraries 
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

In [146]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Standardize the features 
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# Create a logistic regression classifier 
clf = LogisticRegression()

# Train the classifier on the training data
clf.fit(X_train, y_train)

# Make predictions on the testing data
y_pred = clf.predict(X_test)

# Evaluate the model's performance
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)

print(f'Accuracy: {accuracy}')
print(f'Classification Report:\n{report}')

Accuracy: 0.5905349794238683
Classification Report:
              precision    recall  f1-score   support

           0       0.61      0.60      0.61       253
           1       0.57      0.58      0.57       233

    accuracy                           0.59       486
   macro avg       0.59      0.59      0.59       486
weighted avg       0.59      0.59      0.59       486



## **Conclusion**
#### **What we can see is a base betting models that uses the most common NBA statistical data combined with the market's view (ie: moneyline) results in 59% accuracy.**


### **Next steps to improve the model's accuracy**
* Improved feature engineering, including rolling averages for 3, 5, and 10 games.
* Integration of ELO ratings
* Incorporation of team's player statistics
* Integration of historical match-up data